File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
private:
    bool isOpen(char c)
    {
        return c == '(' || c=='{' || c=='[';
    }
    
    char getClose(char c)
    {
        if(c =='(')
            return ')';
        if(c =='[')
            return ']';
        if(c =='{')
            return '}';
        return '\0';
    }
public:
    bool isValid(string s) {
        
        stack<char> st;
        int size = s.size();
        
        int i =0;
        
        while(i<size)
        {
           if(isOpen(s[i]))
               st.push(getClose(s[i]));
           else if(st.empty() || s[i] != st.top())
               return false;
           else
               st.pop();
           i++; 
        }
        return st.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    
    
    
    def isValid(self, s: str) -> bool:
        
        st = []
        openBracs = ['(', '{', '[']
        mp = { ')' : '(',  '}' : '{', ']':'[' }
        
        for ch in s:
            
            if ch in openBracs:
                st.append(ch)
            elif len(st) > 0 and st[len(st) - 1] ==  mp[ch]:
                st.pop()
            else:
                return False
            
        return len(st) == 0
        